python/nx-cugraph/nx_cugraph/algorithms/link_analysis/pagerank_alg.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import cupy as cp
+import networkx as nx
+import numpy as np
+import pylibcugraph as plc
+
+from nx_cugraph.convert import _to_graph
+from nx_cugraph.utils import index_dtype, networkx_algorithm
+
+__all__ = ["pagerank"]
+
+
+@networkx_algorithm
+def pagerank(
+    G,
+    alpha=0.85,
+    personalization=None,
+    max_iter=100,
+    tol=1.0e-6,
+    nstart=None,
+    weight="weight",
+    dangling=None,
+):
+    """`dangling` parameter is not supported."""
+    # Use np.float32 dtype everywhere for performance. In the future, it may
+    # be nice to use whatever dtype the user provides (such as np.float64) and
+    # let PLC "do the right thing" to make sure dtypes of inputs are compatible.
+    G = _to_graph(G, weight, 1, np.float32)
+    if (N := len(G)) == 0:
+        return {}
+    if nstart is not None:
+        nstart = G._dict_to_nodearray(nstart, 0, dtype=np.float32)
+        if (total := nstart.sum()) == 0:
+            raise ZeroDivisionError
+        nstart /= total
+    if personalization is not None:
+        personalization = G._dict_to_nodearray(personalization, 0, dtype=np.float32)
+        if (total := personalization.sum()) == 0:
+            raise ZeroDivisionError
+        personalization /= total
+    if dangling is not None:
+        dangling = G._dict_to_nodearray(dangling, 0)  # Check validity
+        if dangling.sum() == 0:
+            raise ZeroDivisionError
+        if (G._out_degrees_array() == 0).any():
+            raise NotImplementedError("custom dangling weights is not supported")
+    if max_iter <= 0:
+        raise nx.PowerIterationFailedConvergence(max_iter)
+    kwargs = {
+        "resource_handle": plc.ResourceHandle(),
+        "graph": G._get_plc_graph(weight, 1, np.float32, store_transposed=True),
+        "precomputed_vertex_out_weight_vertices": None,
+        "precomputed_vertex_out_weight_sums": None,
+        "initial_guess_vertices": None
+        if nstart is None
+        else cp.arange(N, dtype=index_dtype),  # Why is this necessary?
```

Comment:
I find it strange that we need to provide node indices here (and 12 lines below). I think other places in PLC assume `0, 1, ..., N-1` if indices aren't given.